File: app/src/main/java/com/proxy/shadowsocksr/util/ScreenUtil.java
package com.proxy.shadowsocksr.util;

import android.content.Context;
import android.graphics.Point;
import android.os.Build;
import android.view.Display;
import android.view.WindowManager;

public class ScreenUtil
{
    public static int dp2px(Context context, int dp)
    {
        float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dp * scale + 0.5f);
    }

    public static Point getNavigationBarSize(Context context)
    {
        Point appUsableSize = getAppUsableScreenSize(context);
        Point realScreenSize = getRealScreenSize(context);

        // navigation bar on the right
        if (appUsableSize.x < realScreenSize.x)
        {
            return new Point(realScreenSize.x - appUsableSize.x, appUsableSize.y);
        }

        // navigation bar at the bottom
        if (appUsableSize.y < realScreenSize.y)
        {
            return new Point(appUsableSize.x, realScreenSize.y - appUsableSize.y);
        }

        // navigation bar is not present
        return new Point();
    }

    public static Point getAppUsableScreenSize(Context context)
    {
        WindowManager windowManager = (WindowManager) context
                .getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        return size;
    }

    public static Point getRealScreenSize(Context context)
    {
        WindowManager windowManager = (WindowManager) context
                .getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Point size = new Point();

        if (Build.VERSION.SDK_INT >= 17)
        {
            display.getRealSize(size);
        }
        else if (Build.VERSION.SDK_INT >= 14)
        {
            try
            {
                size.x = (Integer) Display.class.getMethod("getRawWidth").invoke(display);
                size.y = (Integer) Display.class.getMethod("getRawHeight").invoke(display);
            }
            catch (Exception ignored)
            {
            }
        }

        return size;
    }
}

File: app/src/main/java/com/proxy/shadowsocksr/util/ScreenUtil.kt
package com.proxy.shadowsocksr.util

import android.content.Context
import android.graphics.Point
import android.os.Build
import android.view.Display
import android.view.WindowManager

class ScreenUtil
{
    companion object
    {
        @JvmStatic fun dp2px(context: Context, dp: Int): Int
        {
            val scale = context.resources.displayMetrics.density
            return (dp * scale + 0.5f).toInt()
        }

//        @JvmStatic fun getNavigationBarSize(context: Context): Point
//        {
//            val appUsableSize = getAppUsableScreenSize(context)
//            val realScreenSize = getRealScreenSize(context)
//
//            // navigation bar on the right
//            if (appUsableSize.x < realScreenSize.x)
//            {
//                return Point(realScreenSize.x - appUsableSize.x, appUsableSize.y)
//            }
//
//            // navigation bar at the bottom
//            if (appUsableSize.y < realScreenSize.y)
//            {
//                return Point(appUsableSize.x, realScreenSize.y - appUsableSize.y)
//            }
//
//            // navigation bar is not present
//            return Point()
//        }
//
//        @JvmStatic fun getAppUsableScreenSize(context: Context): Point
//        {
//            val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
//            val display = windowManager.defaultDisplay
//            val size = Point()
//            display.getSize(size)
//            return size
//        }
//
//        @JvmStatic fun getRealScreenSize(context: Context): Point
//        {
//            val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
//            val display = windowManager.defaultDisplay
//            val size = Point()
//
//            if (Build.VERSION.SDK_INT >= 17)
//            {
//                display.getRealSize(size)
//            }
//            else if (Build.VERSION.SDK_INT >= 14)
//            {
//                try
//                {
//                    size.x = Display::class.java.getMethod("getRawWidth").invoke(display) as Int
//                    size.y = Display::class.java.getMethod("getRawHeight").invoke(display) as Int
//                }
//                catch (ignored: Exception)
//                {
//                }
//
//            }
//
//            return size
//        }
    }
}
